src/main/java/com/github/manolo8/darkbot/utils/login/LoginUtils.java

```diff
@@ -47,10 +50,69 @@ public static LoginData performUserLogin() {
         Popups.showMessageSync("Login", pane, panel::setDialog);
 
         LoginData loginData = panel.getResult();
-        if (loginData.getPreloaderUrl() == null || loginData.getParams() == null) System.exit(0);
+        if (loginData.getPreloaderUrl() == null || loginData.getParams() == null) {
+            System.out.println("Closed login panel, exited without logging in");
+            System.exit(0);
+        }
         return loginData;
     }
 
+    public static LoginData performAutoLogin(StartupParams params) {
+        String username = params.get(StartupParams.PropertyKey.USERNAME);
+        String password = params.get(StartupParams.PropertyKey.PASSWORD);
+        char[] masterPw = params.getMasterPassword();
+
+        if (username == null || (masterPw == null && (password == null || password.isEmpty()))) {
+            System.err.println("Credentials file requires username and either a password or a master password");
+            System.exit(-1);
+        }
+
+        password = getPassword(username, password, masterPw);
+
+        LoginData loginData = new LoginData();
+        loginData.setCredentials(username, password);
+
+        try {
+            System.out.println("Auto logging in (1/2)");
+            usernameLogin(loginData);
+            System.out.println("Loading spacemap (2/2)");
+            findPreloader(loginData);
+        } catch (WrongCredentialsException e) {
+            System.err.println("Wrong credentials, check your username and password");
+        }
+
+        if (loginData.getPreloaderUrl() == null || loginData.getParams() == null) {
+            System.err.println("Could not find preloader url or parameters");
+            System.exit(-1);
+        }
+        return loginData;
+    }
+
+    private static String getPassword(String username, String password, char[] masterPassword) {
+        if (password == null || password.isEmpty()) {
+            Credentials credentials = loadCredentials();
+            try {
+                credentials.decrypt(masterPassword);
+            } catch (Exception e) {
+                System.err.println("Couldn't retreive logins, check your startup properties file");
+                e.printStackTrace();
+            }
+
+            Credentials.User user = credentials.getUsers()
+                    .stream()
+                    .filter(usr -> username.equals(usr.u))
+                    .findFirst()
+                    .orElse(null);
+            if (user == null) {
+                System.err.println("Couldn't find the provided user in the saved logins. Make sure the username and master password are correct.");
+                password = null;
+            } else {
+                password = user.p;
+            }
+        }
+        return password;
+    }
```

Comment:
What's the point of "getPassword" if the requested parameter is the password?
```suggestion
    private static String getPassword(String username, char[] masterPassword) {
        Credentials credentials = loadCredentials();
        try {
            credentials.decrypt(masterPassword);
        } catch (Exception e) {
            System.err.println("Couldn't retreive logins, check your startup properties file");
            e.printStackTrace();
            return null;
        }
        return credentials.getUsers().stream()
                .filter(usr -> username.equals(usr.u))
                .findFirst().map(u -> u.p).orElse(null);
    }
```
Simplify the function, return null if any issue occurs getting the user.
Possibly even consider `orElseThrow` and throw an exception for missing user or something, but that may be a just complicating it for no good reason